tensorpack/train/trainers.py

```diff
@@ -415,21 +441,24 @@ def _setup_graph(self, input, get_cost_fn, get_opt_fn):
             opt = get_opt_fn()
             self.train_op = opt.apply_gradients(grads, name='train_op')
 
-        def broadcast(self):
-            logger.info("Running horovod broadcast ...")
-            # the op will be created later in initialize()
-            self.trainer._broadcast_op.run()
-
-        cb = CallbackFactory(trigger=broadcast).set_chief_only(False)
+        cb = CallbackFactory(
+            before_train=self.broadcast,
```

Comment:
What does this 'before_train' do? Should that not be conditioned on the BROADCAST_EVERY_EPOCH flag as well?

Reply:
We need to run broadcast op anyway before train.